src/main/scala/d2dadapter/CRCGenerator.scala

```diff
@@ -0,0 +1,132 @@
+import chisel3.util._
+import chisel3.util.Decoupled
+import edu.berkeley.cs.ucie.digital.d2dadapter.CRC16Lookup
+
+/** Generates the CRC of data using the polynomial x^16 + x^15 + x^2 + 1. The
+  * message must be formatted according to the UCIe1.1 spec, and the message
+  * will be consumed one byte per clock cycle, starting from the MSB to the LSB.
+  * The calculated 2 byte CRC will be available on the crc ReadyValid3IO
+  * interface in the raw, un-reversed format as indicated in the UCIe1.1 spec
+  * (i.e. CRC[15] is the MSB of the crc interface's data bits).
+  * @param width
+  *   The message size in bits (must be whole number of bytes). UCIe1.1 spec
+  *   will use 1024-bit messages, with shorter messages padded with 0s in MSB
+  *   side.
+  * @groupdesc Signals
+  *   The input and output controls of the CRCGenerator module
+  */
+
+class CRCGenerator(width: Int) extends Module { //
+  val io = IO(new Bundle {
+
+    /** Resets the CRCGenerator, and sets the default message ready state high
+      * and crc valid state low. CRC bits will be cleared to 0.
+      * @group Signals
+      */
+    val reset = Input(Bool())
+
+    /** ReadyValid3IO interface to allow consumer to transfer message bits to
+      * the CRCGenerator. Message bits are latched when message ready and
+      * message valid are high for the same clock cycle, and then the
+      * CRCGenerator is locked into the calculation loop.
+      *
+      * The CRCGenerator will not accept new messages during the calculation
+      * loop, nor reflect any changes that occur to the message data bits after
+      * the message was latched on the ready/valid transfer.
+      *
+      * The CRCGenerator will only accept a new message after the CRC data has
+      * been transfered through the CRC ReadyValid3IO interface.
+      * @group Signals
+      */
+    val message = Flipped(Decoupled(Bits(width.W)))
+
+    /** ReadyValid3IO interface to allow CRCGenerator to transfer crc bits to
+      * the consumer. CRC will be cleared to 0 and invalid on reset.
+      *
+      * Once message bits have been transfered, the CRCGenerator will enter the
+      * calculation loop, consuming the message data one byte each clock cycle
+      * until all bytes have been consumed. The CRC data bits on the interface
+      * willl then become valid with the crc valid signal set high.
+      *
+      * The CRC data will remain valid and the CRCGenerator will remain in an
+      * idle state until the consumer initiates a transfer of the CRC data by
+      * signaling CRC ready high. After this transfer, the CRC data will become
+      * invalid and the CRC Generator will signal message ready to accept a new
+      * message.
+      * @group Signals
+      */
+    val crc = Decoupled(Bits(16.W))
+  })
+
+  // Output data registers
+  val crc0 = RegInit(Bits(8.W), 0.U)
+  val crc1 = RegInit(Bits(8.W), 0.U)
+
+  // Output signal registers
+  val message_ready = RegInit(Bool(), true.B)
```

Comment:
nit: camelcase?

Reply:
I think I was told to use snake case by @rahulk29, but I can switch it to camelcase if that is recommended.